keel-core/src/main/kotlin/com/netflix/spinnaker/keel/persistence/ArtifactRepository.kt

```diff
@@ -250,6 +250,15 @@ interface ArtifactRepository : PeriodicallyCheckedRepository<DeliveryArtifact> {
     version: String
   ): ArtifactSummaryInEnvironment?
 
+  /**
+   * Get a map of the versions and status for an environmnet
+   */
+  fun getVersionInfoInEnvironment(
+    deliveryConfig: DeliveryConfig,
+    environmentName: String,
+    artifact: DeliveryArtifact
+  ): List<StatusInfoForArtifactInEnvironment>
+
```

Comment:
We query this information _a lot_ of times, and previously we were just looking it up for each individual version. The main change I made in this PR is to pull info for all versions at once and then pass that into subsequent function calls so they look it up from the list we pulled all at once. This allows us to eliminate a bunch of calls to the db.

Reply:
what happens if a new version comes in in the meantime?

Reply:
we will get that version on the next load of the environments UI. The perf gain is definitely worth it here - i'd rather the env load be snappy